### PR TITLE
rails new: add x86_64-linux and aarch64-linux platforms by default

### PR DESCRIPTION
**Pull request by byroot**
Ref: https://github.com/rails/rails/pull/47479

The vast majority of apps will be deployed on Linux, so might as well initialize the Gemfile.lock with these platforms. This will save users from having to do it themselves when they push to GitHub Actions or similar platforms.

FYI: @rafaelfranca @zzak @rubys @dhh

### CODE REVIEW

**Comment by matthewd** [comment]
First thought is I'm not sure how strongly this assumption holds for Windows-based developers (and possibly significant because they're also IMO more likely to end up wanting more... platform-specific gems).

Other concern is that aarch64 is wise to automatically support in principle, but if someone's not actively deploying there, having Bundler refuse to resolve various reasonable-sounding gems because they don't have a platform-compatible release seems like a not-great experience.

**Comment by byroot** [comment]
> I'm not sure how strongly this assumption holds for Windows-based developers

I was under the impression that Windows was extremely rarely used for deployments, only for development.

> Other concern is that aarch64

Yes, I'm a bit on the fence about the `aarch64-linux`, maybe we'd only add that one if you run `rails new` on `macOS/arm64`?

File: railties/lib/rails/generators/app_base.rb [hunk 1]
```diff
@@ -644,7 +644,17 @@ def target_rails_prerelease(self_command = "new")
...
       def run_bundle
-        bundle_command("install", "BUNDLE_IGNORE_MESSAGES" => "1") if bundle_install?
+        if bundle_install?
+          bundle_command("install", "BUNDLE_IGNORE_MESSAGES" => "1")
+
+          # The vast majority of Rails apps will be deployed on `x86_64-linux`.
+          platform = ["--add-platform=x86_64-linux"]
+
+          # Users that develop on M1 mac may use docker and would need `aarch64-linux` as well.
+          platform << "--add-platform=aarch64-linux" if RUBY_PLATFORM.start_with?("arm64")
```
Comment (byroot) [inline railties/lib/rails/generators/app_base.rb:654]: @matthewd I now only add `aarch64-linux` if you run `rails new` from `macOS/M1`.

If you feel strongly about Windows, I can similarly not add `x86_64-linux` if you are running on Windows.

**Comment by matthewd** [comment]
I think Windows is used rarely in general... I have no particular opinion on its frequency for deployment, just a loose theory that given it's so rare _even_ in development, that use in development might indicate being a Windows shop where you Windows all the things. 

I think it's probably fine, though I don't imagine Bundler gives a nice "this would resolve if you dropped platform &lt;X>" error on failure.

**Comment by byroot** [comment]
>  given it's so rare even in development, that use in development might indicate being a Windows shop where you Windows all the things.

My impression was that it was people who don't want to buy an expensive Apple machine and don't want to work in a Linux VM on run a full Linux Desktop, but that still deploy to Heroku or other common deploy targets.

But I assume that with WSL2, actual Windows setups will get rarer and rarer. It's all conjuncture though.

**Comment by rubys** [comment]
My experience is tainted by the fact that I'm the author of a book that gets you started with Rails, and now working for an employer with a free tier - but combined I would say that Windows is used more than you would expect, and that adding `x86_64-linux` is required to deploy to fly.io, and finally that I've not seen a windows user have problems deploying to fly.io due to not being able to find `x86_64-linux` compatible gems (which is essentially what is being done now for deployment, given that BUNDLE_DEPLOYMENT isn't set).

None of these are show stoppers for me as I can always add `x86_64-linux` if not found at `fly launch` time.

**Comment by rubys** [comment]
> The vast majority of apps will be deployed on Linux

This probably should be:

> The vast majority of apps *that are deployed to a cloud or datacenter* will be deployed on Linux.

I have no way to gauge how common that is.  But thinking through the social implications, any platforms that Rails adds will encourage gem providers to support those platforms.  For the rest: places like stack overflow will tell users how to remove platforms that they don't use.  Overall, I'm not seeing huge downsides.

**Comment by byroot** [comment]
Alright, I messed my test, but I'll merge on green.

**Comment by rubys** [comment]
If you create a new rails project on an M1 mac, add these platforms, and then push to github with a github actions defined that does a bundle install without excluding test you will see the following error:

```
Downloading debug-1.7.1 revealed dependencies not in the API or the lockfile
(irb (>= 1.5.0), reline (>= 0.3.1)).
Either installing with `--full-index` or running `bundle update debug` should
fix the problem.
Error: The process '/opt/hostedtoolcache/Ruby/3.2.1/x64/bin/bundle' failed with exit code 34
```

I fixed this in dockerfile-rails:

https://github.com/rails/rails/pull/47506/commits/49aec2c95a4037bc5189c6c2417345e53f462b34#diff-eadf4526a334fb54e822c9880879c7f8989f4fb7b7225d4e088eb463b2df820fR17

... but a more general fix is needed.

**Comment by byroot** [comment]
> you will see the following error:

Yeah, I noticed this one as well and was planning to investigate. Looks like a bundler bug (or feature I don't get) to me.

`debug` is listed as not having any dependencies in the `Gemfile.lock`, not sure why yet, I suspect it's because both dependencies are satisfied by default gems, but not 100% sure.

**Comment by byroot** [comment]
So it's indeed a bug: https://github.com/rubygems/rubygems/issues/6082#issuecomment-1329756343

**Comment by byroot** [comment]
Unfortunately I don't see much we can do on our side :/

Downgrading to `debug 1.7.0` kinda fix this, so we may use that as a temporary workaround, but...

**Comment by dhh** [comment]
Let's make sure we just keep main running with no errors, whatever the intermediary steps needed for that.

**Comment by byroot** [comment]
Yeah, first step is to have a CI to make sure the fix work. What worries me is that if I understand the issue right, it happens whenever the `debug` version matches the one shipped with the current Ruby, so downgrading might fix one version but not the other :/

**Comment by rubys** [comment]
> Unfortunately I don't see much we can do on our side :/

I'm continuing to experiment for the best fix, but the solution may be "if these need to be declared in the Gemfile, then let's declare it and move on".  There is precedent for doing so:

https://github.com/rails/rails/blob/a124668f46c51bb6218e5485b238683677ec1e04/actionmailer/actionmailer.gemspec#L42-L44

**Pull request merged**